rcl_logging_spdlog/src/rcl_logging_spdlog.cpp

```diff
@@ -0,0 +1,174 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cerrno>
+#include <memory>
+#include <mutex>
+#include <inttypes.h>
+
+#include <rcutils/allocator.h>
+#include <rcutils/filesystem.h>
+#include <rcutils/get_env.h>
+#include <rcutils/process.h>
+#include <rcutils/snprintf.h>
+#include <rcutils/time.h>
+
+#include "spdlog/spdlog.h"
+#include "spdlog/sinks/basic_file_sink.h"
+
+#include "rcl_logging_spdlog/logging_interface.h"
+
+#define RC_LOGGING_RET_OK                          (0)
+#define RC_LOGGING_RET_ERROR                       (2)
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static std::mutex g_logger_mutex;
+static std::shared_ptr<spdlog::logger> g_root_logger = nullptr;
+
+/* These are defined here to match the severity levels in rcl. They provide a consistent way for external logger
+    implementations to map between the incoming integer severity from ROS to the concept of DEBUG, INFO, WARN, ERROR,
+    and FATAL*/
+enum RC_LOGGING_LOG_SEVERITY
+{
+  RC_LOGGING_SEVERITY_UNSET = 0,  ///< The unset log level
+  RC_LOGGING_SEVERITY_DEBUG = 10,  ///< The debug log level
+  RC_LOGGING_SEVERITY_INFO = 20,  ///< The info log level
+  RC_LOGGING_SEVERITY_WARN = 30,  ///< The warn log level
+  RC_LOGGING_SEVERITY_ERROR = 40,  ///< The error log level
+  RC_LOGGING_SEVERITY_FATAL = 50,  ///< The fatal log level
+};
+
+static spdlog::level::level_enum map_external_log_level_to_library_level(int external_level)
+{
+  spdlog::level::level_enum level;
+
+  // map to the next highest level of severity
+  if (external_level <= RC_LOGGING_SEVERITY_DEBUG) {
+    level = spdlog::level::level_enum::debug;
+  } else if (external_level <= RC_LOGGING_SEVERITY_INFO) {
+    level = spdlog::level::level_enum::info;
+  } else if (external_level <= RC_LOGGING_SEVERITY_WARN) {
+    level = spdlog::level::level_enum::warn;
+  } else if (external_level <= RC_LOGGING_SEVERITY_ERROR) {
+    level = spdlog::level::level_enum::err;
+  } else if (external_level <= RC_LOGGING_SEVERITY_FATAL) {
+    level = spdlog::level::level_enum::critical;
+  }
+  return level;
+}
+
+rcl_logging_ret_t rcl_logging_external_initialize(const char * config_file, rcutils_allocator_t allocator)
+{
+  std::lock_guard<std::mutex> lk(g_logger_mutex);
+  // It is possible for this to get called more than once in a process (some of
+  // the tests do this implicitly by calling rclcpp::init more than once).
+  // If the logger is already setup, don't do anything.
+  if (g_root_logger != nullptr) {
+    return RC_LOGGING_RET_OK;
+  }
+
+  bool config_file_provided = (nullptr != config_file) && (config_file[0] != '\0');
+  if (config_file_provided) {
+    // TODO(clalancette): implement support for an external configuration file.
+    return RC_LOGGING_RET_ERROR;
```

Comment:
Does it make sense to set an error message with `RCUTILS_SET_ERROR_MSG` before returning? Same for other error cases below.

Reply:
Yeah, that makes sense.  I just didn't know about RCUTILS_SET_ERROR_MSG.  Will fix.